lib/ride/services/audio.dart

```diff
@@ -0,0 +1,820 @@
+
+/// The distances before the crossings when a speed advisory instruction should be played.
+const List<int> speedAdvisoryDistances = [300, 100];
+
+/// The distances until an instruction is played even if the distance is smaller than the speedAdvisoryDistances.
+const List<int> speedAdvisoryMinDistances = [200, 50];
+
+/// The minimal distance before the crossing when a speed advisory instruction should be played. (Except wait for green)
+const int speedAdvisoryMinDistance = 50;
+
+const double predictionQualityThreshold = 0.85;
+
+class Audio {
+  /// An instance for text-to-speach.
+  FlutterTts? ftts;
+
+  /// The tracking service instance.
+  Tracking? tracking;
+
+  /// The ride service instance.
+  Ride? ride;
+
+  /// The settings service instance.
+  Settings? settings;
+
+  /// The positioning service instance.
+  Positioning? positioning;
+
+  /// Whether the audio service is initialized.
+  bool initialized = false;
+
+  /// A map that holds information about the last recommendation to check the difference when a new recommendation is received.
+  Map<String, Object> lastRecommendation = {};
+
+  /// The current route.
+  Route? currentRoute;
+
+  /// The last signal group id for which a wait for green info was played.
+  String? didStartWaitForGreenInfoTimerForSg;
+
+  /// The wait for green timer that is used to time the wait for green instruction.
+  Timer? waitForGreenTimer;
+
+  /// The current state of the speed advisory instruction.
+  int currentSpeedAdvisoryInstructionState = 0;
+
+  /// The last signal group id.
+  int lastSignalGroupId = -1;
+
+  /// The last 11 values of the user speed.
+  List<double> lastSpeedValues = [];
+
+  /// An instance of the audio session.
+  AudioSession? audioSession;
+
+  /// The last prediction from the prediction service.
+  Prediction? lastPrediction;
+
+  /// Constructor.
+  Audio() {
+    settings = getIt<Settings>();
+    settings!.addListener(_processSettingsUpdates);
+
+    if (settings!.audioSpeedAdvisoryInstructionsEnabled) {
+      initialized = true;
+      _init();
+    }
+  }
+
+  /// Init the audio service.
+  Future<void> _init() async {
+    ride ??= getIt<Ride>();
+    ride!.addListener(_processRideUpdates);
+    positioning ??= getIt<Positioning>();
+    positioning!.addListener(_processPositioningUpdates);
+  }
+
+  /// Clean up the audio instructions feature.
+  Future<void> cleanUp() async {
+    ride = null;
+    positioning?.removeListener(_processPositioningUpdates);
+    positioning = null;
+    await resetFTTS();
+    // Deactivate the audio session to allow other audio to play.
+    audioSession?.setActive(false);
+    audioSession = null;
+    lastRecommendation.clear();
+  }
+
+  /// Reset the complete audio service.
+  Future<void> reset() async {
+    settings?.removeListener(_processSettingsUpdates);
+    settings = null;
+    waitForGreenTimer?.cancel();
+    waitForGreenTimer = null;
+    didStartWaitForGreenInfoTimerForSg = null;
+    await cleanUp();
+  }
```

Comment:
We are currently having `cleanUp()`, `reset()` and `resetFTTS()`. We should simplify this and make the names more clearer. Maybe we can even make some methods private. Right now there is lots of room for misuse/errors when accidentally calling the wrong method.

Reply:
Same for `_initializeTTS`, `init` and `initAudioService`

Reply:
In general, this class is really hard to read/understand, because it does a lot of different things. Is it maybe possible to split some stuff into another class? (Like the FTTS parts into a separate class that manages its state on its own and that has its own init and reset method.)